libdevcrypto/LibSnark.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ This file is part of cpp-ethereum.
+
+ cpp-ethereum is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ cpp-ethereum is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file LibSnark.cpp
+ */
+
+#include <libdevcrypto/LibSnark.h>
+
+#include <algebra/curves/alt_bn128/alt_bn128_g1.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_g2.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_pairing.hpp>
+#include <algebra/curves/alt_bn128/alt_bn128_pp.hpp>
+
+#include <libdevcrypto/Exceptions.h>
+
+#include <libdevcore/CommonIO.h>
+#include <libdevcore/Log.h>
+
+#include <mutex>
+
+using namespace std;
+using namespace dev;
+using namespace dev::crypto;
+
+namespace
+{
+
+std::once_flag initLibSnarkFlag;
+
+void initLibSnark()
+{
+	call_once(initLibSnarkFlag,
+		[](){
+			cnote << "calling libff::alt_bn128_pp::init_public_params().";
+			libff::alt_bn128_pp::init_public_params();
+		});
+}
+
+libff::bigint<libff::alt_bn128_q_limbs> toLibsnarkBigint(h256 const& _x)
+{
+	libff::bigint<libff::alt_bn128_q_limbs> x;
+	for (unsigned i = 0; i < 4; i++)
+		for (unsigned j = 0; j < 8; j++)
+			x.data[3 - i] |= uint64_t(_x[i * 8 + j]) << (8 * (7 - j));
+	return x;
+}
+
+h256 fromLibsnarkBigint(libff::bigint<libff::alt_bn128_q_limbs> _x)
+{
+	h256 x;
+	for (unsigned i = 0; i < 4; i++)
+		for (unsigned j = 0; j < 8; j++)
+			x[i * 8 + j] = uint8_t(uint64_t(_x.data[3 - i]) >> (8 * (7 - j)));
+	return x;
+}
+
+libff::alt_bn128_Fq decodeFqElement(dev::bytesConstRef _data)
+{
+	// h256::AlignLeft ensures that the h256 is zero-filled on the right if _data
+	// is too short.
+	h256 xbin(_data, h256::AlignLeft);
+	if (u256(xbin) >= u256(fromLibsnarkBigint(libff::alt_bn128_Fq::mod)))
+		BOOST_THROW_EXCEPTION(InvalidEncoding());
+	return toLibsnarkBigint(xbin);
+}
+
+libff::alt_bn128_G1 decodePointG1(dev::bytesConstRef _data)
+{
+	libff::alt_bn128_Fq X = decodeFqElement(_data.cropped(0));
```

Comment:
No, now I see that the code is supposed to work even when `_data` is shorter than 64 bytes.